From 79cafcc1d19a4814d3e74b4830a2a4d34f5e99cc Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Mon, 12 Oct 2015 16:23:03 -0700
Subject: [PATCH 21/26] irq-gic-v3-its: Setup PCI requester ID.
Patch-mainline: Submitted, https://lkml.org/lkml/2015/10/8/890
Reference: fate#319484

DO NOT UPSTREAM.  For Linux v4.3, a different patch
is needed and has been submitted.

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Matthias Brugger <mbrugger@suse.com>

---
 drivers/irqchip/irq-gic-v3-its.c |   10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

--- a/drivers/irqchip/irq-gic-v3-its.c
+++ b/drivers/irqchip/irq-gic-v3-its.c
@@ -1258,7 +1258,6 @@ static int its_alloc_device_irq(struct i
 
 struct its_pci_alias {
 	struct pci_dev	*pdev;
-	u32		dev_id;
 	u32		count;
 };
 
@@ -1276,7 +1275,6 @@ static int its_get_pci_alias(struct pci_
 {
 	struct its_pci_alias *dev_alias = data;
 
-	dev_alias->dev_id = alias;
 	if (pdev != dev_alias->pdev)
 		dev_alias->count += its_pci_msi_vec_count(dev_alias->pdev);
 
@@ -1290,6 +1288,7 @@ static int its_msi_prepare(struct irq_do
 	struct its_node *its;
 	struct its_device *its_dev;
 	struct its_pci_alias dev_alias;
+	u32 dev_id;
 
 	if (!dev_is_pci(dev))
 		return -EINVAL;
@@ -1299,20 +1298,21 @@ static int its_msi_prepare(struct irq_do
 	dev_alias.count = nvec;
 
 	pci_for_each_dma_alias(pdev, its_get_pci_alias, &dev_alias);
+	dev_id = pci_msi_domain_get_msi_rid(domain, pdev);
 	its = domain->parent->host_data;
 
-	its_dev = its_find_device(its, dev_alias.dev_id);
+	its_dev = its_find_device(its, dev_id);
 	if (its_dev) {
 		/*
 		 * We already have seen this ID, probably through
 		 * another alias (PCI bridge of some sort). No need to
 		 * create the device.
 		 */
-		dev_dbg(dev, "Reusing ITT for devID %x\n", dev_alias.dev_id);
+		dev_dbg(dev, "Reusing ITT for devID %x\n", dev_id);
 		goto out;
 	}
 
-	its_dev = its_create_device(its, dev_alias.dev_id, dev_alias.count);
+	its_dev = its_create_device(its, dev_id, dev_alias.count);
 	if (!its_dev)
 		return -ENOMEM;
 
